steward/src/cork.rs

```diff
@@ -0,0 +1,107 @@
+use crate::{
+    cellars::{self, uniswapv3, CellarId},
+    error::{Error, ErrorKind},
+    prelude::APP,
+    somm_send,
+};
+use abscissa_core::{
+    tracing::log::{debug, error},
+    Application,
+};
+use deep_space::{Coin, Contact};
+use gravity_bridge::gravity_proto::cosmos_sdk_proto::cosmos::base::abci::v1beta1::TxResponse;
+use somm_proto::cork::Cork;
+use std::time::Duration;
+use steward_proto::steward::{
+    self,
+    submit_request::ContractCallData::{self, *},
+    SubmitRequest, SubmitResponse,
+};
+use tonic::async_trait;
+
+pub struct CorkHandler;
+
+#[async_trait]
+impl steward::contract_call_server::ContractCall for CorkHandler {
+    async fn submit(
+        &self,
+        request: tonic::Request<SubmitRequest>,
+    ) -> Result<tonic::Response<SubmitResponse>, tonic::Status> {
+        debug!("received contract call request");
+        tokio::spawn(async move {
+            let request = request.get_ref();
+            let cork = match build_cork(request).await {
+                Ok(c) => c,
+                Err(err) => {
+                    error!("{}", err);
+                    return;
+                }
+            };
+            debug!("cork: {:?}", cork);
+
+            if let Err(err) = send_cork(cork).await {
```

Comment:
Before sending, it would be useful to add a check to see if this is even a cellar that has been added by governance. The chain will also catch this error, but we can catch it too before submitting a vote by retrieving the values from the chain.